Opengl tests cleanup

ronen opened the pull request:
DRY up the OpenGL test suite by adding helpers `Testing::fill` and `Testing::check_result`, which are themselves wrappers around `Buffer.for_each_element`.

Also along the way cleaned up some of the tests themselves.  In particular, `shifted_domains.cpp` didn't exit with 1 in case of failure.

cc: @estollnitz @shoaibkamil

estollnitz commented:
LGTM.

Like I've said privately, I'd love to see `Halide::Runtime::Buffer<T>` acquire methods similar to `Testing::fill` and `Testing::check_result`, and see them used throughout the Halide tests. However, judging from the complexity of the `Buffer::for_each_element` implementation, it would require quite a bit of code to support functions of any dimensionality. That's a task for another time.

steven-johnson commented on test/opengl/conv_select.cpp:
```diff
@@ -4,6 +4,8 @@
 #include "Halide.h"
 #include <stdio.h>
 
+#include "./testing.h"
```
Why the ./ prefix here: is "testing.h" not adequate?

ronen replied:
A habit of mine -- intends to hint to a reader that the .h file is in the local directory rather than elsewhere in the include path.  But not important.  I switched them to "testing.h"

steven-johnson commented on test/opengl/testing.h:
```diff
@@ -0,0 +1,93 @@
+
+    // fill 3-dimension buffer
+    template <typename T>
+        void fill(Halide::Buffer<T> &buf, std::function<T(int x, int y, int c)> f)
```
Indentation and bracing doesn't follow Halide conventions: This should really be

namespace Testing {

template<typename T>
void fill (...) {
   ...
}

ronen replied:
Sorry wasn't paying attention to that... and didn't realize there's a .clang-format file.  I ran clang-format on everything in test/opengl

steven-johnson replied:
The .clang-format file isn't little-used, but is probably a good bet :-)

steven-johnson commented on test/opengl/testing.h:
```diff
@@ -0,0 +1,93 @@
+                public: static void vector(const std::vector<T> &v) {
+                    for (size_t i=0; i<v.size(); i++) {
+                        if (i > 0) std::cerr << ",";
+                        std::cerr << v[i]+0; // need to add 0 to get output -- compiler bug??
```
Depends entirely on what type T is. Where do you see it?

ronen replied:
Ah yeah.  Fails when T is `uint8_t` since that's not actually an integer it's a char, and so gets output as a (generally garbage) ascii character.   Simpler fix then is to use `+v[i]` so that the unary + operator will do the promotion to integer.   Switched to that and updated the comment.

steven-johnson commented on test/opengl/testing.h:
```diff
@@ -0,0 +1,93 @@
+                    }
+                }
+            };
+            try {
```
Many Halide configs are built with exceptions disabled; the try/catch needs to be ifdef'ed appropriately. See Error.h/.cpp for an example.

ronen replied:
Yes, on my platform Halide is built with exceptions disabled.  But in the Makefile that's only for building the Halide runtime; the tests don't have exceptions disabled.  And the Makefile has [[L133]](https://github.com/halide/Halide/blob/master/Makefile#L133) `CXX_FLAGS += -funwind-tables` with a comment `"on some hosts... Without -funwind-tables, libHalide.so fails to propagate exceptions and causes a test failure."` 

All of which seems to tell me it should be OK to use exceptions in the tests.  Is there some config in which the tests have exceptions disabled?

steven-johnson commented:
> I'd love to see Halide::Runtime::Buffer<T> acquire methods similar to Testing::fill and Testing::check_result, and see them used throughout the Halide tests

I think test-specific utilities of this sort make more sense to be in a separate library, built on top of the existing Buffer::for_each_element(), etc; this looks like a very good start. I suspect we'll want to enhance this and move it into tools/ for general use.

abadams commented:
Please see #1849 for a generic implementation of the fill method.

ronen commented:
@steven-johnson Thanks for all the comments.  I've addressed them all, though see my reply re exceptions in the tests.

@abadams Way cool, thanks.  I've merged the latest master, and switched over to `Buffer::fill`

steven-johnson commented on test/opengl/lut.cpp:
```diff
@@ -46,30 +51,14 @@ int test_lut1d() {
+                case 2: return (x > 3) ? 8.0f : 1.0f;
+                default: return std::numeric_limits<float>::infinity();
+            } }))
+        return 1;
```
Nit: IMHO this would be clearer with explicit braces around the if clause. (I realize we don't come close to a uniform style on this in Halide, but generally, prefer explicit braces for all if/else/for blocks, even if single-statement.)

ronen replied:
Yep.  Added braces.

steven-johnson commented on test/opengl/sum_reduction.cpp:
```diff
@@ -33,22 +31,15 @@ int main() {
-    }
+            return temp / 10.0f * 255.0f;
+        },
+                                      1e-3))
```
Youch, here's an example of where clang-format followed the rules we gave it correctly, and yet gave a weird and eyeball-hurting result.

ronen replied:
In some way, once past the initial shock of this, it's actually not that unreasonable.  Having (optional) parameters after a lambda block is an awkward and hard-to-parse thing in any case, and  this certainly makes it clear what's going on.

...and having written that, it becomes clear that I should get rid of the awkward code.  

So I moved the tolerance parameter before the block.  And made it optional via an overloaded two-parameter template that calls the three-parameter version with 0 as the second argument.

That eases the eyeball pain :)
